### ISSUE TITLE
Speed issues [enhancement]

### ISSUE THREAD

**Issue by reg4in** [feross/standard#43]
First of all, I like the style of standard and I use it in all projects I started lately.

Buuuut, it always takes several seconds to test a single file in my test directory even if it only requires small files. My test file is about 150 LOC, the source is in this case even smaller.

Maybe this is an issue not with standard it self but with one of it's dependencies ... this is really interrupting my workflow.

standard@2.7.3
Intel Core i5, 4 Gigs RAM, Ubuntu 14.04

**Comment by feross** [comment]
After investigating, there are two main places where time is spent:

1. ~500ms. Locating the `eslint` and `jscs` binaries with the `npm run which-eslint` and `npm run which-jscs` commands. We locate the binaries in parallel. See https://github.com/feross/standard/pull/30 for why we need to do this.

2. ~400ms running the linters. We run the linters in parallel.

We could speed things up a lot by using the `eslint` and `jscs` JS APIs directly, instead of relying on the command line programs. This would eliminate step 1 entirely, and eliminate a `child_process.exec` call in step 2.

PR welcome – I don't have time for this right at the moment.

**Comment by reg4in** [comment]
I'll dig into that when I have some free time, maybe I can do something ... :+1:

**Comment by Flet** [comment]
eh, first try at using eslint directly... WIP
https://github.com/Flet/standard/commit/6bb49bac0ca0b70a47a154b7cddfaf408c324a59

**Comment by feross** [comment]
@Flet Good start. :+1: I say we hold off on merging this until we've eliminated `jscs` (#45), so we don't waste effort converting that too.

**Comment by reg4in** [comment]
@feross 
You're the boss, but as far as I looked in to that I think that actually these two can work side by side quiet harmonously.

@Flet
Are you going to finish that? Or should I begin to work on it too?

**Comment by Flet** [comment]
@reg4in yeah probably... but maybe it makes sense to create an `eslint-only` branch in this repo and continue work collaboratively?

Once the eslint rules from #45 are in, I'd guess we would want remove jscs and gain the speed benefits of using eslint directly. The branch could focus on that end state.

Thoughts, @feross? Or should we redirect our efforts to creating the missing eslint rules instead? :)

**Comment by reg4in** [comment]
I'd totally be in :+1:

**Comment by feross** [comment]
@Flet Either approach is fine. I just added you as a [collaborator](https://github.com/feross/webtorrent/blob/master/CONTRIBUTING.md) on the repo.

**Comment by feross** [comment]
Thanks for the first attempt code, @Flet – it was very helpful.

I just pushed a branch (speed), which improves performance of `standard` by 2-4x! This branch does both items discussed in my first comment:

1. remove the need for `npm run which-eslint`.
2. remove the `child_process.exec` calls and use the node api of `eslint` directly.

I also remove `jscs`, which we're close to doing anyway (https://github.com/feross/standard/issues/45).

Linting feross/standard:

```
$ time standard
standard  2.21s user 0.36s system 161% cpu 1.596 total
$ time ./bin/cmd.js
./bin/cmd.js  0.61s user 0.06s system 105% cpu 0.630 total
```

Linting npm/npm:

```
$ time standard
standard  4.63s user 0.44s system 175% cpu 2.895 total
$ time ./bin/cmd.js
./bin/cmd.js  2.18s user 0.16s system 114% cpu 2.052 total
```

**Comment by Flet** [comment]
Sweet! I was just thinking of returning to this with the eslint rule movement :) Amazing what a few dollars does! :moneybag: 

Taking a gander at the branch now.

**Comment by Flet** [comment]
Wow, so much cruft gone now! Also should be easier to use `standard` programmatically now that it takes a callback.

**Comment by Flet** [comment]
I take full blame for the sync stuff :sob:

**Comment by feross** [comment]
Hah, no worries. I'm still doing a bit more cleanup to make the api nicer for programmatic use :)

**Comment by feross** [comment]
Okay! New, cleaned up node.js apis are available on the [`speed` branch](https://github.com/feross/standard/tree/speed). Really happy with how they turned out!

Deleted around 200 lines total, increased speed 2-4x, and it's way clearer how everything works now!

**Comment by reg4in** [comment]
nice one!

**Comment by millermedeiros** [comment]
instead of using `npm run which-eslint` you could use the [npm-run](https://github.com/timoxley/npm-run) lib. the slow part is not the `which` command, but spawning the npm executable itself.

**Comment by feross** [comment]
Ah, handy! Wish I knew about `npm-run` before.

In any case, we've removed `npm run` and `child_process.exec` in favor of `require('eslint')` which requires less code and fewer hacks. :) This is currently on the `speed` branch and will become the next version of `standard` once https://github.com/eslint/eslint/issues/1619 is fixed.

**Comment by feross** [comment]
This is released as 3.3.0.

**Comment by reg4in** [comment]
nice! good job!
